cmds/wifi/wifi.go

```diff
@@ -38,34 +38,59 @@ func init() {
 	}
 }
 
+func generateConfig(a []string) (conf []byte, err error) {
+	// format of a: [essid, pass, id]
+	switch {
+	case len(a) == 3:
+		conf = []byte(fmt.Sprintf(eap, a[0], a[2], a[1]))
+	case len(a) == 2:
+		conf, err = passphrase.Run(a[0], a[1])
+		if err != nil {
+			return nil, fmt.Errorf("essid: %v, pass: %v : %v", a[0], a[1], err)
+		}
+	case len(a) == 1:
+		conf = []byte(fmt.Sprintf(nopassphrase, a[0]))
+	default:
+		flag.Usage()
+		os.Exit(1)
+	}
+	return
+}
+
 func main() {
+	go startServer()
+
+	// Service
 	var (
 		iface = flag.String("i", "wlan0", "interface to use")
-		essid string
 		conf  []byte
+		err   error
 	)
 
 	flag.Parse()
 	a := flag.Args()
 
-	switch {
-	case len(a) == 3:
-		essid = a[0]
-		conf = []byte(fmt.Sprintf(eap, essid, a[2], a[1]))
-	case len(a) == 2:
-		essid = a[0]
-		pass := a[1]
-		o, err := passphrase.Run(essid, pass)
-		if err != nil {
-			log.Fatalf("%v %v: %v", essid, pass, err)
+	if len(a) == 0 {
+		// Experimental Part
+		// if len(a) = 0, can use the web interface to get user's input
+		msg := <-ServerToServiceChan
+		a = append(a, msg.essid)
+		if msg.pass != "" {
+			a = append(a, msg.pass)
 		}
-		conf = o
-	case len(a) == 1:
-		essid = a[0]
-		conf = []byte(fmt.Sprintf(nopassphrase, essid))
-	default:
-		flag.Usage()
-		os.Exit(1)
+		if msg.id != "" {
+			a = append(a, msg.id)
+		}
+		stubMsg := ServiceToServerMessage{
+			essid: msg.essid,
+		}
+		ServiceToServerChan <- stubMsg
+		_ = <-ServerToServiceChan // (Experimental) So we can see the result of the page load
+	}
+
+	conf, err = generateConfig(a)
+	if err != nil {
+		log.Fatalf("error: %v", err)
 	}
 
 	if err := ioutil.WriteFile("/tmp/wifi.conf", conf, 0444); err != nil {
```

Comment:
so we only see the wpa_supplicant output if exec.Command got the error? What is the supplicant exits normally? Shouldn't we see its output if it is not empty?

Reply:
Sorry, I don't quite understand this comment. The only error that can come out from generateConfig is if wpa_passphrase has an error or if invalid argument is passed in. Any of those two will cause the to exit which is similar to what we had before. Those lines are just standard error checking.